lib/protobuf/field/field_hash.rb

```diff
@@ -0,0 +1,97 @@
+module Protobuf
+  module Field
+    class FieldHash < Hash
+
+      ##
+      # Attributes
+      #
+
+      attr_reader :field, :key_field, :value_field
+
+      ##
+      # Constructor
+      #
+
+      def initialize(field)
+        @field = field
+        @key_field = field.type_class.get_field(:key)
+        @value_field = field.type_class.get_field(:value)
+      end
+
+      ##
+      # Public Instance Methods
+      #
+
+      def []=(key, val)
+        super(normalize_key(key), normalize_val(val)) unless val.nil?
```

Comment:
Now throws an exception. Also made more changes to `FieldArray` so that its internal state cannot be corrupted with other methods/features inherited from `Array`.